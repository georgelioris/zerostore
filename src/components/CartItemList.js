import React from "react";
import Icon from "@material-ui/core/Icon";

const subTotal = items => cartItem => {
  return getProperty(items)(cartItem, "price") * cartItem.quantity;
};

const getProperty = items => (cartItem, prop) => {
  const key = cartItem.id;
  return items.hasOwnProperty(key) ? items[key][prop] : "Item Removed";
};

const CartItem = ({
  items,
  cartItem,
  onClickInc,
  onClickDec,
  onClickRemove
}) => (
  <div className="card white darken-1" key={cartItem.id}>
    <div className="card-content black-text">
      <span className="card-title">
        {getProperty(items)(cartItem, "title")}
        <div className="remove">
          <span onClick={onClickRemove}>
            <Icon>clear</Icon>
          </span>
        </div>
      </span>

      <div className="action">
        <span
          className={
            "btn-floating white " + (cartItem.quantity === 1 ? "disabled" : "")
          }
          onClick={onClickDec}
        >
          <i className="material-icons black-text">remove</i>
        </span>
        <span className="quantity">{cartItem.quantity}</span>
        <span className="btn-floating white" onClick={onClickInc}>
          <i className="small material-icons black-text">add</i>
        </span>
        <br />
        <div className="subTotal">
          Subtotal: {subTotal(items)(cartItem)}
          {` `}$
        </div>
      </div>
    </div>
  </div>
);

const CartItemList = ({ cartItems, items, IncQuant, DecQuant, RemoveItem }) => (
  <div className="cartList">
    <h4>My Cart</h4>
    <div className="total">
      Total:
      <span>
        {cartItems.reduce((acc, cartItem) => {
          return acc + subTotal(items)(cartItem);
        }, 0)}
        {` `}$
      </span>
    </div>
    {cartItems.map(cartItem => {
      return (
        <CartItem
          key={cartItem.id}
          cartItem={cartItem}
          items={items}
          onClickInc={() => IncQuant(cartItem)}
          onClickDec={() => DecQuant(cartItem)}
          onClickRemove={() => RemoveItem(cartItem)}
        />
      );
    })}
  </div>
);

export default CartItemList;
