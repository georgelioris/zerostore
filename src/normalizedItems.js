import ItemImg from "./images/swshoer.png";

export const defaultItems = {
  1: {
    id: 1,
    title: "winter body",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 110,
    img: ItemImg,
    category: "winter",
    available: true
  },
  2: {
    id: 2,
    title: "Adidas",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 80,
    img: ItemImg,
    category: "new",
    available: true
  },
  3: {
    id: 3,
    title: "Vans",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 120,
    img: ItemImg,
    category: "on sale",
    available: true
  },
  4: {
    id: 4,
    title: "White",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 260,
    img: ItemImg,
    category: "on sale",
    available: true
  },
  5: {
    id: 5,
    title: "Cropped-shoe",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 160,
    img: ItemImg,
    category: "summer",
    available: true
  },
  6: {
    id: 6,
    title: "Blues",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 90,
    img: ItemImg,
    category: "summer",
    available: true
  },
  7: {
    id: 7,
    title: "Black",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 390,
    img: ItemImg,
    category: "winter",
    available: true
  },
  8: {
    id: 8,
    title: "Another One",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 40,
    img: ItemImg,
    category: "new",
    available: true
  },
  9: {
    id: 9,
    title: "Orange",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 190,
    img: ItemImg,
    category: "new",
    available: true
  },
  10: {
    id: 10,
    title: "Model One",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 55,
    img: ItemImg,
    category: "new",
    available: true
  }
};
