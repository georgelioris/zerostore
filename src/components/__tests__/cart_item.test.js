import React from "react";
import CartItemList from "../CartItemList";
import { unmountComponentAtNode } from "react-dom";
import { act, create } from "react-test-renderer";
import { defaultItems } from "../../normalizedItems";

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

// Mock state and actions
const mock = jest.fn();
const items = defaultItems;
const cartItems = [{ id: 1, quantity: 1 }, { id: 4, quantity: 2 }];
const props = {
  items: items,
  cartItems: cartItems,
  incQuant: mock,
  decQuant: mock,
  removeFromCart: mock
};

// Tests
describe("CartItemList component", () => {
  const container = create(<CartItemList {...props} />);
  it("Matches the snapshot", () => {
    expect(container.toJSON()).toMatchSnapshot();
  });
  it("Should dispatch incQuant/decQaunt on button click", () => {
    // Click all incQuant and decQaunt buttons
    act(() => {
      container.root
        .findAllByProps({
          className: "btn-floating white"
        })
        .map(i => i.props.onClick());
    });

    expect(props.incQuant).toHaveBeenCalledTimes(2);
    expect(props.decQuant).toHaveBeenCalledTimes(2);
    expect(props.incQuant).toHaveBeenNthCalledWith(1, { id: 1, quantity: 1 });
    expect(props.incQuant).toHaveBeenNthCalledWith(2, { id: 4, quantity: 2 });
    expect(props.decQuant).toHaveBeenNthCalledWith(1, { id: 1, quantity: 1 });
    expect(props.decQuant).toHaveBeenNthCalledWith(2, { id: 4, quantity: 2 });
  });
  it("Should dispatch removeFromCart on button click", () => {
    // Click all removeFromCart buttons
    act(() => {
      container.root.findAllByProps({
        className: "remove"
      });
    });

    expect(props.removeFromCart).toHaveBeenCalledTimes(2);
    expect(props.removeFromCart).toHaveBeenNthCalledWith(1, {
      id: 1,
      quantity: 1
    });
    expect(props.removeFromCart).toHaveBeenNthCalledWith(2, {
      id: 4,
      quantity: 2
    });
  });
});
