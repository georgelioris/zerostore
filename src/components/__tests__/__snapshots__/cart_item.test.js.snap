// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemList component Matches the snapshot 1`] = `
<div
  className="cartList"
>
  <h4>
    My Cart
  </h4>
  <div
    className="card white"
  >
    <div
      className="card-content black-text row"
    >
      <div
        className="col s2"
      >
        <div
          className="image-component circle"
          style={
            Object {
              "background": "url(swshoer.png)",
              "height": "120px",
              "width": "120px",
            }
          }
        />
      </div>
      <div
        className="col s10"
      >
        <span
          className="card-title"
        >
          Winter body
          <div
            className="remove"
          >
            <span
              onClick={[Function]}
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                clear
              </span>
            </span>
          </div>
        </span>
        <div
          className="action"
        >
          <span
            className="btn-floating white disabled"
            onClick={[Function]}
          >
            <i
              className="material-icons black-text"
            >
              remove
            </i>
          </span>
          <span
            className="quantity"
          >
            1
          </span>
          <span
            className="btn-floating white"
            onClick={[Function]}
          >
            <i
              className="small material-icons black-text"
            >
              add
            </i>
          </span>
          <br />
          <div
            className="subTotal"
          >
            Subtotal: 110$
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="card white"
  >
    <div
      className="card-content black-text row"
    >
      <div
        className="col s2"
      >
        <div
          className="image-component circle"
          style={
            Object {
              "background": "url(swshoer.png)",
              "height": "120px",
              "width": "120px",
            }
          }
        />
      </div>
      <div
        className="col s10"
      >
        <span
          className="card-title"
        >
          White
          <div
            className="remove"
          >
            <span
              onClick={[Function]}
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                clear
              </span>
            </span>
          </div>
        </span>
        <div
          className="action"
        >
          <span
            className="btn-floating white "
            onClick={[Function]}
          >
            <i
              className="material-icons black-text"
            >
              remove
            </i>
          </span>
          <span
            className="quantity"
          >
            2
          </span>
          <span
            className="btn-floating white"
            onClick={[Function]}
          >
            <i
              className="small material-icons black-text"
            >
              add
            </i>
          </span>
          <br />
          <div
            className="subTotal"
          >
            Subtotal: 520$
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="total"
  >
    <span>
      Total: 630 $
    </span>
  </div>
</div>
`;
